### ISSUE TITLE
blockdevices are found and then immediately disappear

### ISSUE THREAD

**Issue by edwmurph** [openebs/cstor-operators#417]
<details>
<summary>I'm installing via flux:</summary>

```
apiVersion: source.toolkit.fluxcd.io/v1beta1
kind: HelmRepository
metadata:
  name: openebs-cstor
  namespace: flux-system
spec:
  interval: 10m
  url: https://openebs.github.io/cstor-operators
---
apiVersion: helm.toolkit.fluxcd.io/v2beta1
kind: HelmRelease
metadata:
  name: cstor
  namespace: openebs
spec:
  interval: 10m
  chart:
    spec:
      chart: cstor
      interval: 5m
      sourceRef:
        kind: HelmRepository
        name: openebs-cstor
        namespace: flux-system
  values:
    csiNode:
      kubeletDir: "/var/snap/microk8s/common/var/lib/kubelet/"
    openebs-ndm:
      ndm:
        filters:
          includePaths: "/dev/mmcblk0p3"
      featureGates:
        UseOSDisk:
          enabled: true
```
</details>

<details>
<summary>onto a microk8s cluster with 3 4gb mem raspberry pi nodes</summary>

```
300s 6:50:30 ~/code/personal/personal⟩ kubectl get nodes                                                                                     ± master
NAME   STATUS   ROLES    AGE   VERSION
rpi2   Ready    <none>   15h   v1.23.3-2+0d2db09fa6fbbb
rpi3   Ready    <none>   18h   v1.23.3-2+0d2db09fa6fbbb
rpi1   Ready    <none>   15h   v1.23.3-2+0d2db09fa6fbbb
```

```
ubuntu@rpi3:~$ sudo microk8s.kubectl version
Client Version: version.Info{Major:"1", Minor:"23+", GitVersion:"v1.23.3-2+0d2db09fa6fbbb", GitCommit:"0d2db09fa6fbbbf8e7fd3e4c9d9e1c4b29f66481", GitTreeState:"clean", BuildDate:"2022-01-26T22:26:56Z", GoVersion:"go1.17.6", Compiler:"gc", Platform:"linux/arm64"}
Server Version: version.Info{Major:"1", Minor:"23+", GitVersion:"v1.23.3-2+0d2db09fa6fbbb", GitCommit:"0d2db09fa6fbbbf8e7fd3e4c9d9e1c4b29f66481", GitTreeState:"clean", BuildDate:"2022-01-26T22:17:11Z", GoVersion:"go1.17.6", Compiler:"gc", Platform:"linux/arm64"}
ubuntu@rpi3:~$ sudo microk8s status
microk8s is running
high-availability: yes
  datastore master nodes: 192.168.86.47:19001 192.168.86.26:19001 192.168.86.48:19001
  datastore standby nodes: none
addons:
  enabled:
    dns                  # CoreDNS
    ha-cluster           # Configure high availability on the current node
    rbac                 # Role-Based Access Control for authorisation
```
</details>

<details>
<summary>each pi has a 128GB microsd card with 3 partitions: ubuntu boot (256M), OS filesystem (13.7G) and a 104.3G unmounted primary ext4 partition with no filesystem:</summary>

```
ubuntu@rpi3:~$ sudo fdisk -l /dev/mmcblk0
Disk /dev/mmcblk0: 119.9 GiB, 127865454592 bytes, 249737216 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: dos
Disk identifier: 0xf66f0719

Device         Boot    Start       End   Sectors   Size Id Type
/dev/mmcblk0p1 *        2048    526335    524288   256M  c W95 FAT32 (LBA)
/dev/mmcblk0p2        526336  29296639  28770304  13.7G 83 Linux
/dev/mmcblk0p3      29296640 248047615 218750976 104.3G 83 Linux
```
</details>

<details>
<summary>all the openebs pods start up correctly:</summary>

```
1s 6:53:44 ~/code/personal/personal⟩ kc get pods --all-namespaces                                                                             ± master
NAMESPACE     NAME                                          READY   STATUS    RESTARTS      AGE
kube-system   coredns-64c6478b6c-qhtst                      1/1     Running   0             15h
kube-system   calico-node-fhs4d                             1/1     Running   0             15h
kube-system   calico-node-wwkmp                             1/1     Running   0             15h
kube-system   calico-kube-controllers-6966456d6b-8h6rt      1/1     Running   0             18h
flux-system   notification-controller-77766855f7-ml97h      1/1     Running   0             15h
flux-system   source-controller-74766f477b-nhj8f            1/1     Running   0             15h
kube-system   calico-node-t8wbm                             1/1     Running   1 (90m ago)   15h
flux-system   kustomize-controller-856c6868cd-cb6hd         1/1     Running   2 (87m ago)   15h
flux-system   helm-controller-688d5444b-t5dx9               1/1     Running   2 (87m ago)   15h
openebs       cstor-csi-node-wg9zp                          2/2     Running   0             66s
openebs       cstor-openebs-ndm-62jww                       1/1     Running   0             66s
openebs       cstor-csi-node-nwhlh                          2/2     Running   0             66s
openebs       cstor-csi-node-vxhtf                          2/2     Running   0             66s
openebs       cstor-cvc-operator-6f57946444-ln747           1/1     Running   0             66s
openebs       cstor-openebs-ndm-xlc2z                       1/1     Running   0             66s
openebs       cstor-admission-server-77c48bb8c-7xsh8        1/1     Running   0             66s
openebs       cstor-openebs-ndm-7dcbx                       1/1     Running   0             66s
openebs       cstor-cspc-operator-cf87fdc89-p9wm6           1/1     Running   0             66s
openebs       cstor-csi-controller-0                        6/6     Running   0             66s
openebs       cstor-openebs-ndm-operator-55b6c8846b-mn2j2   1/1     Running   0             66s
```
</details>

<details>
<summary>and all the block devices are initially found if i watch the resources while openebs is being installed:</summary>

```
725s 6:51:31 ~/code/personal/personal⟩ kc get bd --all-namespaces --watch                                                                  ± ⇡1 master
NAMESPACE   NAME                                           NODENAME   SIZE           CLAIMSTATE   STATUS   AGE
openebs     blockdevice-638111d10bab619c876b59c88ae99d95   rpi1       112000499712   Unclaimed    Active   1s
openebs     blockdevice-638111d10bab619c876b59c88ae99d95   rpi3       112000499712   Unclaimed    Active   2s
openebs     blockdevice-638111d10bab619c876b59c88ae99d95   rpi2       112000499712   Unclaimed    Active   5s
```
</details>

<details>
<summary>but then the blockdevices seem to immediately overwrite eachother because the only blockdevice that remains is the last one that is found:</summary>

```
247s 6:55:39 ~/code/personal/personal⟩ kc get bd --all-namespaces --watch                                                                     ± master
NAMESPACE   NAME                                           NODENAME   SIZE           CLAIMSTATE   STATUS   AGE
openebs     blockdevice-638111d10bab619c876b59c88ae99d95   rpi2       112000499712   Unclaimed    Active   2m39s
```
</details>

openebs with cstor seems to be working really well for my use case so far other than this hurdle so i would greatly appreciate any help!

**Comment by akhilerm** [comment]
@edwmurph Were the micro SD cards cloned from  a single source. It seems like the partition UUID for `mmcblk0p3` is same on all the 3 nodes. Can you confirm it?

If yes, then you can use fdisk to change the partition UUID on 2 nodes and then it should work fine.

**Comment by edwmurph** [comment]
thanks for the quick response!

> It seems like the partition UUID for mmcblk0p3 is same on all the 3 nodes. Can you confirm it?

yes that's right and appears to be causing the issue. all three disks had a `/dev/mmcblk0p3` primary partition which each generate the same block device name in k8s `blockdevice-638111d10bab619c876b59c88ae99d95` even though they're on different disks/nodes

<details>
<summary>i updated the <code>includePaths</code> to <code>"/dev/mmcblk0p3,/dev/mmcblk0p4"</code> and then used fdisk to change one disk to use the 4th primary partition slot <code>/dev/mmcblk0p4</code> and then pretty shortly after ndm started reporting 2 block devices:</summary>

```
9s 7:17:18 ~/code/personal/personal⟩ kc get bd --all-namespaces --watch                                    ± master
NAMESPACE   NAME                                           NODENAME   SIZE           CLAIMSTATE   STATUS   AGE
openebs     blockdevice-935fdbf41be557aafdbabbd0da4c2014   rpi2       112865574912   Unclaimed    Active   10s
openebs     blockdevice-638111d10bab619c876b59c88ae99d95   rpi3       112865574912   Unclaimed    Active   7s
```
</details>

however i still can't figure out how to get block devices tracked by ndm for all 3 of my sd cards. the sd cards im using require the boot os in the first partition and the OS filesystem in the 2nd partition so i only have 2 partition names (`/dev/mmcblk0p3` and `/dev/mmcblk0p4`) to work with on each sd card and i'd like all 3 raspberry pi sd cards to be used in the storage pool

**Comment by akhilerm** [comment]
You dont need to change anything in `includePaths`.   How we identify the disks/partitions is by using the UUID present on the disk.

Can you give the output of `udevadm info /dev/mmcblk0p3`  and `udevadm info /dev/mmcblk0p4` from all the 3 nodes. Need to check the `PART_ENTRY_UUID` field which we use for identification.

If we can have unique `PART_ENTRY_UUID` on the partition on each nodes, then all cards from all nodes can be used without any issue.

Also, Were the micro SD cards cloned from a single source ?

**Comment by edwmurph** [comment]
> You dont need to change anything in includePaths. How we identify the disks/partitions is by using the UUID present on the disk.

<details>
<summary>ah ya i confirmed on my end that the <code>includePaths</code> filter isn't needed. but it appears i need <code>featureGates.UseOSDisk.enabled=true</code> which is also finding my the boot partition so i ended up also updating the <code>excludePaths</code> to ignore the boot partition:</summary>

```
    csiNode:
      kubeletDir: "/var/snap/microk8s/common/var/lib/kubelet/"
    openebs-ndm:
      ndm:
        filters:
          excludePaths: "/dev/mmcblk0p1,/dev/mmcblk0p2,loop,fd0,sr0,/dev/ram,/dev/dm-,/dev/md,/dev/rbd,/dev/zd"
      featureGates:
        UseOSDisk:
          enabled: true
```
</details>

> Also, Were the micro SD cards cloned from a single source ?

yes i initialized them all with Raspberry Pi Imager from my macbook

<hr />

<details>
<summary><code>ubuntu@rpi1:~$ sudo fdisk /dev/mmcblk0 -l</code></summary>

```
Disk /dev/mmcblk0: 119.9 GiB, 127865454592 bytes, 249737216 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: dos
Disk identifier: 0xf66f0719

Device         Boot    Start       End   Sectors   Size Id Type
/dev/mmcblk0p1 *        2048    526335    524288   256M  c W95 FAT32 (LBA)
/dev/mmcblk0p2        526336  29296639  28770304  13.7G 83 Linux
/dev/mmcblk0p3      29296640 249737215 220440576 105.1G 83 Linux
```
</details>

<details>
<summary><code>ubuntu@rpi1:~$ udevadm info /dev/mmcblk0p3</code></summary>

```
P: /devices/platform/emmc2bus/fe340000.emmc2/mmc_host/mmc0/mmc0:aaaa/block/mmcblk0/mmcblk0p3
N: mmcblk0p3
L: 0
S: disk/by-id/mmc-SN128_0x88a2c4ea-part3
S: disk/by-path/platform-fe340000.emmc2-part3
S: disk/by-partuuid/f66f0719-03
S: disk/by-uuid/bb84cb38-8a47-4b73-b8bd-384a786942f6
E: DEVPATH=/devices/platform/emmc2bus/fe340000.emmc2/mmc_host/mmc0/mmc0:aaaa/block/mmcblk0/mmcblk0p3
E: DEVNAME=/dev/mmcblk0p3
E: DEVTYPE=partition
E: PARTN=3
E: MAJOR=179
E: MINOR=3
E: SUBSYSTEM=block
E: USEC_INITIALIZED=43951612396
E: ID_SCSI=1
E: ID_NAME=SN128
E: ID_SERIAL=0x88a2c4ea
E: ID_PATH=platform-fe340000.emmc2
E: ID_PATH_TAG=platform-fe340000_emmc2
E: ID_PART_TABLE_UUID=f66f0719
E: ID_PART_TABLE_TYPE=dos
E: ID_DRIVE_FLASH_SD=1
E: ID_DRIVE_MEDIA_FLASH_SD=1
E: ID_FS_UUID=bb84cb38-8a47-4b73-b8bd-384a786942f6
E: ID_FS_UUID_ENC=bb84cb38-8a47-4b73-b8bd-384a786942f6
E: ID_FS_VERSION=1.0
E: ID_FS_TYPE=ext4
E: ID_FS_USAGE=filesystem
E: ID_PART_ENTRY_SCHEME=dos
E: ID_PART_ENTRY_UUID=f66f0719-03
E: ID_PART_ENTRY_TYPE=0x83
E: ID_PART_ENTRY_NUMBER=3
E: ID_PART_ENTRY_OFFSET=29296640
E: ID_PART_ENTRY_SIZE=220440576
E: ID_PART_ENTRY_DISK=179:0
E: DEVLINKS=/dev/disk/by-id/mmc-SN128_0x88a2c4ea-part3 /dev/disk/by-path/platform-fe340000.emmc2-part3 /dev/disk/by-partuuid/f66f0719-03 /dev/disk/by-uuid/bb84cb38-8a47-4b73-b8bd-384a786942f6
E: TAGS=:systemd:
```
</details>

<details>
<summary><code>ubuntu@rpi1:~$ udevadm info /dev/mmcblk0p4</code></summary>

```
Unknown device "/dev/mmcblk0p4": No such file or directory
```
</details>

<hr />

<details>
<summary><code>ubuntu@rpi2:~$ sudo fdisk /dev/mmcblk0 -l</code></summary>

```
Disk /dev/mmcblk0: 119.9 GiB, 127865454592 bytes, 249737216 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: dos
Disk identifier: 0xf66f0719

Device         Boot    Start       End   Sectors   Size Id Type
/dev/mmcblk0p1 *        2048    526335    524288   256M  c W95 FAT32 (LBA)
/dev/mmcblk0p2        526336  29296639  28770304  13.7G 83 Linux
/dev/mmcblk0p4      29296640 249737215 220440576 105.1G 83 Linux
```
</details>

<details>
<summary><code>ubuntu@rpi2:~$ udevadm info /dev/mmcblk0p3</code></summary>

```
Unknown device "/dev/mmcblk0p3": No such file or directory
```
</details>

<details>
<summary><code>ubuntu@rpi2:~$ udevadm info /dev/mmcblk0p4</code></summary>

```
P: /devices/platform/emmc2bus/fe340000.emmc2/mmc_host/mmc0/mmc0:aaaa/block/mmcblk0/mmcblk0p4
N: mmcblk0p4
L: 0
S: disk/by-path/platform-fe340000.emmc2-part4
S: disk/by-partuuid/f66f0719-04
S: disk/by-id/mmc-SN128_0x8902c4ea-part4
S: disk/by-uuid/0a6b4fd2-9f87-43c5-9d78-9529e92c60cc
E: DEVPATH=/devices/platform/emmc2bus/fe340000.emmc2/mmc_host/mmc0/mmc0:aaaa/block/mmcblk0/mmcblk0p4
E: DEVNAME=/dev/mmcblk0p4
E: DEVTYPE=partition
E: PARTN=4
E: MAJOR=179
E: MINOR=4
E: SUBSYSTEM=block
E: USEC_INITIALIZED=148121857485
E: ID_SCSI=1
E: ID_NAME=SN128
E: ID_SERIAL=0x8902c4ea
E: ID_PATH=platform-fe340000.emmc2
E: ID_PATH_TAG=platform-fe340000_emmc2
E: ID_PART_TABLE_UUID=f66f0719
E: ID_PART_TABLE_TYPE=dos
E: ID_DRIVE_FLASH_SD=1
E: ID_DRIVE_MEDIA_FLASH_SD=1
E: ID_FS_UUID=0a6b4fd2-9f87-43c5-9d78-9529e92c60cc
E: ID_FS_UUID_ENC=0a6b4fd2-9f87-43c5-9d78-9529e92c60cc
E: ID_FS_VERSION=1.0
E: ID_FS_TYPE=ext4
E: ID_FS_USAGE=filesystem
E: ID_PART_ENTRY_SCHEME=dos
E: ID_PART_ENTRY_UUID=f66f0719-04
E: ID_PART_ENTRY_TYPE=0x83
E: ID_PART_ENTRY_NUMBER=4
E: ID_PART_ENTRY_OFFSET=29296640
E: ID_PART_ENTRY_SIZE=220440576
E: ID_PART_ENTRY_DISK=179:0
E: DEVLINKS=/dev/disk/by-path/platform-fe340000.emmc2-part4 /dev/disk/by-partuuid/f66f0719-04 /dev/disk/by-id/mmc-SN128_0x8902c4ea-part4 /dev/disk/by-uuid/0a6b4fd2-9f87-43c5-9d78-9529e92c60cc
E: TAGS=:systemd:
```
</details>

<hr />

<details>
<summary><code>ubuntu@rpi3:~$ sudo fdisk /dev/mmcblk0 -l</code></summary>

```
Disk /dev/mmcblk0: 119.9 GiB, 127865454592 bytes, 249737216 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: dos
Disk identifier: 0xf66f0719

Device         Boot    Start       End   Sectors   Size Id Type
/dev/mmcblk0p1 *        2048    526335    524288   256M  c W95 FAT32 (LBA)
/dev/mmcblk0p2        526336  29296639  28770304  13.7G 83 Linux
/dev/mmcblk0p3      29296640 249737215 220440576 105.1G 83 Linux
```
</details>

<details>
<summary><code>ubuntu@rpi3:~$ udevadm info /dev/mmcblk0p3</code></summary>

```
P: /devices/platform/emmc2bus/fe340000.emmc2/mmc_host/mmc0/mmc0:aaaa/block/mmcblk0/mmcblk0p3
N: mmcblk0p3
L: 0
S: disk/by-partuuid/f66f0719-03
S: disk/by-uuid/1fa35e3f-b8c2-455e-b2b1-ffb3cde65b0a
S: disk/by-path/platform-fe340000.emmc2-part3
S: disk/by-id/mmc-SN128_0x8a62c4e6-part3
E: DEVPATH=/devices/platform/emmc2bus/fe340000.emmc2/mmc_host/mmc0/mmc0:aaaa/block/mmcblk0/mmcblk0p3
E: DEVNAME=/dev/mmcblk0p3
E: DEVTYPE=partition
E: PARTN=3
E: MAJOR=179
E: MINOR=3
E: SUBSYSTEM=block
E: USEC_INITIALIZED=152732948628
E: ID_SCSI=1
E: ID_NAME=SN128
E: ID_SERIAL=0x8a62c4e6
E: ID_PATH=platform-fe340000.emmc2
E: ID_PATH_TAG=platform-fe340000_emmc2
E: ID_PART_TABLE_UUID=f66f0719
E: ID_PART_TABLE_TYPE=dos
E: ID_DRIVE_FLASH_SD=1
E: ID_DRIVE_MEDIA_FLASH_SD=1
E: ID_FS_UUID=1fa35e3f-b8c2-455e-b2b1-ffb3cde65b0a
E: ID_FS_UUID_ENC=1fa35e3f-b8c2-455e-b2b1-ffb3cde65b0a
E: ID_FS_VERSION=1.0
E: ID_FS_TYPE=ext4
E: ID_FS_USAGE=filesystem
E: ID_PART_ENTRY_SCHEME=dos
E: ID_PART_ENTRY_UUID=f66f0719-03
E: ID_PART_ENTRY_TYPE=0x83
E: ID_PART_ENTRY_NUMBER=3
E: ID_PART_ENTRY_OFFSET=29296640
E: ID_PART_ENTRY_SIZE=220440576
E: ID_PART_ENTRY_DISK=179:0
E: DEVLINKS=/dev/disk/by-partuuid/f66f0719-03 /dev/disk/by-uuid/1fa35e3f-b8c2-455e-b2b1-ffb3cde65b0a /dev/disk/by-path/platform-fe340000.emmc2-part3 /dev/disk/by-id/mmc-SN128_0x8a62c4e6-part3
E: TAGS=:systemd:
```
</details>

<details>
<summary><code>ubuntu@rpi3:~$ udevadm info /dev/mmcblk0p4</code></summary>

```
Unknown device "/dev/mmcblk0p4": No such file or directory
```
</details>

<hr />

> If we can have unique PART_ENTRY_UUID on the partition on each nodes, then all cards from all nodes can be used without any issue.

i just changed the disk identifier for one of the disks via `fdisk` expert commands, then rebooted the raspberry pi and ndm found all 3 disks! now it looks like everything should just work for my use case 🤞

thanks for all your help! i think this issue is resolved and can be closed and hopefully it helps someone else in the future!

the key insight for me was that the disk identifier + partition number combination needs to be unique for ndm to be able to manage it

**Comment by akhilerm** [comment]
great. 

TLDR; 

>  the key insight for me was that the disk identifier + partition number combination needs to be unique for ndm to be able to manage it

yes, actually we use the Partition UUID, which happens to be partition table ID + partition number in DOS scheme (the scheme used in rpie).


There is one better way to exclude your boot partition. In the OS-disk filter, you can specify the mount path of your boot, like `/boot`, and it would exclude on all the nodes.

**Comment by edwmurph** [comment]
> There is one better way to exclude your boot partition. In the OS-disk filter, you can specify the mount path of your boot, like /boot, and it would exclude on all the nodes.

<details>
<summary>i tried this config</summary>

```
    csiNode:
      kubeletDir: "/var/snap/microk8s/common/var/lib/kubelet/"
    openebs-ndm:
      ndm:
        filters:
          excludePaths: "/boot,/dev/mmcblk0p2,loop,fd0,sr0,/dev/ram,/dev/dm-,/dev/md,/dev/rbd,/dev/zd"
      featureGates:
        UseOSDisk:
          enabled: true
```
</details>

<details>
<summary>but ndm still found the boot partition</summary>

```
1s 9:43:21 ~/code/personal/personal⟩ kc describe bd blockdevice-7acbb9e76e0a480798ebbb964ecca2b5 -n flux-system
Name:         blockdevice-7acbb9e76e0a480798ebbb964ecca2b5
Namespace:    flux-system
Labels:       beta.kubernetes.io/arch=arm64
              beta.kubernetes.io/os=linux
              kubernetes.io/arch=arm64
              kubernetes.io/hostname=rpi1
              kubernetes.io/os=linux
              microk8s.io/cluster=true
              ndm.io/blockdevice-type=blockdevice
              ndm.io/managed=true
              node.kubernetes.io/microk8s-controlplane=microk8s-controlplane
              nodename=rpi1
Annotations:  internal.openebs.io/uuid-scheme: gpt
API Version:  openebs.io/v1alpha1
Kind:         BlockDevice
Metadata:
  Creation Timestamp:  2022-02-10T14:50:49Z
  Generation:          1
  Managed Fields:
    API Version:  openebs.io/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:internal.openebs.io/uuid-scheme:
        f:labels:
          .:
          f:beta.kubernetes.io/arch:
          f:beta.kubernetes.io/os:
          f:kubernetes.io/arch:
          f:kubernetes.io/hostname:
          f:kubernetes.io/os:
          f:microk8s.io/cluster:
          f:ndm.io/blockdevice-type:
          f:ndm.io/managed:
          f:node.kubernetes.io/microk8s-controlplane:
          f:nodename:
      f:spec:
        .:
        f:capacity:
          .:
          f:logicalSectorSize:
          f:physicalSectorSize:
          f:storage:
        f:details:
          .:
          f:compliance:
          f:deviceType:
          f:driveType:
          f:firmwareRevision:
          f:hardwareSectorSize:
          f:logicalBlockSize:
          f:model:
          f:physicalBlockSize:
          f:serial:
          f:vendor:
        f:devlinks:
        f:filesystem:
          .:
          f:fsType:
          f:mountPoint:
        f:nodeAttributes:
          .:
          f:nodeName:
        f:partitioned:
        f:path:
      f:status:
        .:
        f:claimState:
        f:state:
    Manager:         ndm
    Operation:       Update
    Time:            2022-02-10T14:50:49Z
  Resource Version:  7022
  Self Link:         /apis/openebs.io/v1alpha1/namespaces/flux-system/blockdevices/blockdevice-7acbb9e76e0a480798ebbb964ecca2b5
  UID:               4fe63b36-5c64-4ef4-8761-7a5a11db8d36
Spec:
  Capacity:
    Logical Sector Size:   512
    Physical Sector Size:  512
    Storage:               268435456
  Details:
    Compliance:
    Device Type:           partition
    Drive Type:            SSD
    Firmware Revision:
    Hardware Sector Size:  512
    Logical Block Size:    512
    Model:
    Physical Block Size:   512
    Serial:
    Vendor:
  Devlinks:
    Kind:  by-id
    Links:
      /dev/disk/by-id/mmc-SN128_0x88a2c4ea-part1
    Kind:  by-path
    Links:
      /dev/disk/by-path/platform-fe340000.emmc2-part1
  Filesystem:
    Fs Type:      vfat
    Mount Point:  /boot/firmware
  Node Attributes:
    Node Name:  rpi1
  Partitioned:  No
  Path:         /dev/mmcblk0p1
Status:
  Claim State:  Unclaimed
  State:        Active
Events:         <none>
```
</details>

<details>
<summary>however when i switch back to these opts, only the correct partitions are found</summary>

```
    csiNode:
      kubeletDir: "/var/snap/microk8s/common/var/lib/kubelet/"
    openebs-ndm:
      ndm:
        filters:
          excludePaths: "/dev/mmcblk0p1,/dev/mmcblk0p2,loop,fd0,sr0,/dev/ram,/dev/dm-,/dev/md,/dev/rbd,/dev/zd"
      featureGates:
        UseOSDisk:
          enabled: true
```
</details>

am i missing something?

**Comment by akhilerm** [comment]
Which setup are you using? Can you give some details about it. Because NDM has more filtering options available. https://github.com/openebs/node-disk-manager/blob/develop/deploy/helm/charts/values.yaml#L42. You need to add `/boot/firmware` into the `osDiskExcludePaths`

**Comment by edwmurph** [comment]
i'm installing via the [openebs-cstor chart](https://github.com/openebs/cstor-operators/blob/HEAD/deploy/helm/charts/README.md) which doesn't seem to mention `osDiskExcludePaths` in its' [configuration section](https://github.com/openebs/cstor-operators/blob/HEAD/deploy/helm/charts/README.md#configuration)?

**Comment by akhilerm** [comment]
Oh. Its not mentioned in the configuration section. But it can be used.  NDM chart is a dependency of cstor chart. so similar to how other values are set, you can set the ndm values also. Ref: https://github.com/openebs/node-disk-manager/blob/develop/deploy/helm/charts/values.yaml#L44

**Comment by edwmurph** [comment]
ah dope thanks ya that's better and worked!

<details>
<summary>for reference, here's how i'm installing now:</summary>

```
apiVersion: helm.toolkit.fluxcd.io/v2beta1
kind: HelmRelease
metadata:
  name: cstor
  namespace: openebs
spec:
  interval: 10m
  chart:
    spec:
      chart: cstor
      interval: 5m
      sourceRef:
        kind: HelmRepository
        name: openebs-cstor
        namespace: flux-system
  values:
    csiNode:
      kubeletDir: "/var/snap/microk8s/common/var/lib/kubelet/"
    openebs-ndm:
      # ndm options:
      # https://github.com/openebs/node-disk-manager/blob/develop/deploy/helm/charts/values.yaml#L44
      ndm:
        filters:
          osDiskExcludePaths: "/,/etc/hosts,/boot,/boot/firmware"
      featureGates:
        UseOSDisk:
          enabled: true
```
</details>

**Comment by akhilerm** [comment]
Great. So if the issue is solved, can we close it?

**Comment by edwmurph** [comment]
yes this is solved thanks for all you're help!
